### PR TITLE
Week05

### PR DESCRIPTION
**Pull request by arunpariyar**
This PR includes  `store.go`, `errors.go` and `clause.go` alongside their respective test files `store_test.go`, `errors_test.go` and `clause_test.go` for the 5th week assignment. 

The clause_test.go file consists of two test functions. 

- `func Test_clause_String(t *testing.T) {..}`
- `func Test_clause_Match(t *testing.T) {..}`

In the `func Test_clause_String(t *testing.T) {..}` table test is used for three scenarios: “matching clauses”, “unmatching clauses” and “empty clauses”. Each test table consists of test name, clauses and the exp string. Each test is run in a sub-test where the output of the `String()` Method  is checked against the expected string which is created using string literals. If the test fails an error is printed to the terminal. 

`func Test_clause_Match(t *testing.T) {..}` similarly makes use of table test to test for three scenarios: “true case”, “false case” and “empty case”. Each test table consists of test name, clauses, models and the expected boolean. Each test is run in a sub-test where the result of the `Match()` method on the clauses is checked against the expected boolean. If the test fails an error is printed to the terminal. 

The `errors_test.go` consists of the following tests functions. 

- `func Test_errors_Error(t *testing.T) {..}`
- `func Test_errors_Table_Not_Found(t *testing.T) {..}`
- `func Test_errors_ErrTableNotFound_Is(t *testing.T) {..}`
- `func Test_errors_Clauses(t *testing.T) {..}`
- `func Test_errors_RowNotFound(t *testing.T) {..}`
- `func Test_errors_errNoRows_Is(t *testing.T) {..}`
- `func Test_errors_isErrNoRows(t *testing.T) {..}`
- `func Test_errors_AsErrNoRows(t *testing.T) {..}`

`func Test_errors_Error(t *testing.T) {..}` doesn’t make use of table test because it's a rather simple implementation that returns a just as string. First the error created is checked through the `IsErrorTableNotFound()` function, if the check is successful the Error method is called on the error and the output is checked against a string literal. Failure in both checks prints errors to the terminal. Checking if an error is `ErrTableNotFound` allows the program to filter invalid error in at the very first step, which makes the task following specific to its’ implementations. 

`func Test_errors_Table_Not_Found(t *testing.T) {..}` is similar to the above, it first checks the error through the `IsErrorTableNotFound()` function and checks the returned boolean for trigger error. It also doesn't make use of table tests. 

`func Test_errors_ErrTableNotFound_Is(t *testing.T) {..}` checks the created test error if it's a `ErrorTableNotFound{}` if false, error message is printed. It also doesn't use table tests. 
 
`func Test_errors_Clauses(t *testing.T) {..}` makes use of table test to check 3 scenarios “not empty”, “empty” and “clause nil”. In each sub-test the `Clauses()` Method is called and the result is further tested by running its `String()` method the last result is checked against the expected value and if false and error is displayed. It makes use of table tests to accommodate the 3 scenarios.

`func Test_errors_RowNotFound(t *testing.T) {..}` simply checks returns of the sample error which consists of tables and clauses. A direct string comparison of the table which is string triggers the error on false case. The clauses on the other hand calls its `String()` method and then compares the string return with the expected value, if false an error is printed.

`func Test_errors_errNoRows_Is(t *testing.T) {..}` similar to the `Is()` for ErrTableNotFound the test creates a sampler error and runs the `Is()` method with the expected type is `errNoRow` the returned boolean value is checked which in a false scenario triggers an error. A simple unit test is used here rather than a table test. 

`func Test_errors_isErrNoRows(t *testing.T) {..}` test for two cases “true case” & “false case” at table test is used here to accommodate both cases easily. In the sub-test the `isErrNoRows()` function is called for sample errors. And if the expected result is not met an error message is returned. The great thing about using table tests here is the ease with which both true and false cases can be tested. 

`func Test_errors_AsErrNoRows(t *testing.T) {..}` similar to the above table test is used for both true and false cases. The sample errors are called on the `AsErrNoRows` function from the two return the earlier that returns the error type is ignored as for this test the main concern is the returned true or false which we validate against our expected values and print error if the values don’t match. Here table test also makes it easy to check for both true and false cases. 

The `store_test.go` is the test file of the three that makes the most use of table-test. The file consists of the following functions. 

- `func Test_store_db(t *testing.T) {..}`
- `func Test_store_All(t *testing.T) {..}`
- `func Test_store_Len(t *testing.T) {..}`
- `func Test_store_Insert(t *testing.T) {..}`
- `func Test_store_Select(t *testing.T) {..}`
- `func assertSameModels(t testing.TB, act Models, exp Models) {..}`

`func Test_store_db(t *testing.T) {..}` is the shortest test of all since this function is created for the one task of returning empty data in case a store data has a nil value. The sample store is thus one with nil data.  The `db()` method is then called on the store and saved on `db`. The length of db is checked to 0 since it is empty. If false an error is printed.

`func Test_store_All(t *testing.T) {..}` makes use of table test to test two scenerios: “store with orders” and “empty store”. In the sub-test the `All()` method is called on the store and two return values the models and error are stored. First if the error is not nil then a test is done to make sure it is of the expected error type if a false error is printed. The error now handled, the program then calls the helper assertSameModels function that checks the two slices of models for us if any mismatch is found an error is then printed. The checking of slice of Models is done multiple times in the program thus a helper function for this task was most feasible. 

`func Test_store_Len(t *testing.T) {..}` makes use of table test to check for two scenarios “empty store” and “store with orders”. In the sub-tes the Len() method is called on the store and returned length and error is stored. First the error is checked to see if it is the expected error if not an error is triggered. Finally the len is checked against the expected value if the result is false an error is triggered. 

`func Test_store_Insert(t *testing.T) {..}` makes use of table test to check for two scenarios of adding just one and then another item to the store. In the sub-test first the inserts method is called and following that we get the database of the store using the `db()` method. The database is then query for the table wanted which returns the models in the table. The returned models are checked to be the same using the assertSameModels helper function if any mismatch is found an error message is triggered. 

`func Test_store_Select(t *testing.T) {..}` makes use of table test to check three scenarios: “store with table and model”, “store with table but no models” and “query with no clause”. In the subtest the `Select()` method is called on the store with the table name and the clauses. The results are then stored. First the error is handled by checking the error strings to show which error caused the problem. If no error is returned the program checks the models against our expected model. If a mismatch is found and error is returned.

`func assertSameModels(t testing.TB, act Models, exp Models) {..}` is a helper function designed to test if two slices of model are the same. Since our data are layers of maps it wasn't very straightforward to just make use of a simple for loop to compare them. What this helper function does is it creates a separate list of keys and values for both the models and compares them to make sure both models are indeed equal. One small step to mention her is that storing the keys inside the loop was straight forward since keys are string by nature however since our values are of type interface they had to be converted to string for which the code below was used
`vc1 := fmt.Sprintf("%s", v)` // this string prints the interface value as string
`v1 = append(v1, vc1)` // the string is then appended to the value slice. 
	After the keys and values are stored in slice they are also sorted by strings using `sort.Strings(v1)` 

My main difficulties with this assignment was really understanding how table tests, sub-tests and helper functions worked. It also took a fair amount of time to understand what the functions were doing so that I had a better idea of the input and output for each function. The way that I found was to just rewrite the function line by line and experiment. Oh and writing that helper function was not easy. 

A few surprises this week were 
- while writing the table test, one thing that I realized was that rather than creating separate values and passing them into the table instead writing the values directly in the table and formatting them properly very much helped in not going to multiple places to see the code and it also made the code look very attractive.
- Helper functions are awesome. 
- It isn’t easy to compare slices of maps that implement maps inside with interfaces. 
- Using the `fmt.Sprintf`  interface values can be converted to strings.

### CODE REVIEW

**Review by markbates — COMMENTED** [review]
Can you please remove the `week05/go.mod` file so that the tests will run?

**Comment by arunpariyar** [comment]
> Can you please remove the `week05/go.mod` file so that the tests will run?

**Comment by arunpariyar** [comment]
`go.mod` is now deleted

File: week05/clause_test.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		name string
+		cls  Clauses
+		mdl  Model
+		exp  bool
+	}{
+		{
+			name: "true case",
+			cls:  Clauses{"name": "Sir Canon Doyle", "job": "Investigator"},
+			mdl:  Model{"name": "Sir Canon Doyle", "job": "Investigator"},
+			exp:  true},
```
Comment (markbates) [inline week05/clause_test.go:53]: Put the closing curly brace on a new line to match the opening one. I found this very confusing. At first I couldn't figure out how the code was even compiling.

```suggestion
		{
			name: "true case",
			cls:  Clauses{"name": "Sir Canon Doyle", "job": "Investigator"},
			mdl:  Model{"name": "Sir Canon Doyle", "job": "Investigator"},
			exp:  true,
		},
```

File: week05/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
+package demo
+
+import (
+	"testing"
+)
+
+func Test_errors_Error(t *testing.T) {
```
Comment (markbates) [inline week05/errors_test.go:7]: Try to use clearer and more descriptive test names.

```suggestion
func Test_IsErrTableNotFound(t *testing.T) {
```

File: week05/errors_test.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	e := ErrTableNotFound{
+		table: "test",
+	}
+	exp := true
+
+	act := e.Is(ErrTableNotFound{})
+	if act != true {
+		t.Fatalf("expected %t got %t", exp, act)
+	}
+}
```
Comment (markbates) [inline week05/errors_test.go:54]: These tests could be collapsed into a table driven test.

File: week05/errors_test.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	for _, tt := range table {
+		t.Run(tt.name, func(t *testing.T) {
+			_, act := AsErrNoRows(tt.e)
+
+			if act != tt.exp {
+				t.Fatalf("%s expected %t got %t", tt.name, tt.exp, act)
+			}
+		})
+	}
+}
```
Comment (markbates) [inline week05/errors_test.go:190]: Again, these tests feel like they could be table driven.

File: week05/store_test.go [hunk 5]
```diff
@@ -0,0 +1,304 @@
+package demo
+
+import (
+	"fmt"
+	"sort"
+	"testing"
+)
+
+func Test_store_db(t *testing.T) {
```
Comment (markbates) [inline week05/store_test.go:9]:
```suggestion
func Test_Store_db(t *testing.T) {
```

File: week05/store_test.go [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+	}
+
+	db := s.db()
+
+	if len(db) != 0 {
+		t.Fatalf("expected 0 got %v", len(db))
+	}
+}
+
+func Test_store_All(t *testing.T) {
```
Comment (markbates) [inline week05/store_test.go:22]:
```suggestion
func Test_Store_All(t *testing.T) {
```

File: week05/store_test.go [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+	sort.Strings(k2)
+	sort.Strings(v1)
+	sort.Strings(v2)
+
+	for i := 0; i < len(k1); i++ {
+		if k1[i] != k2[i] || v1[i] != v2[i] {
+			t.Fatal("models don't match")
+		}
+	}
+}
```
Comment (markbates) [inline week05/store_test.go:304]:
```suggestion
func assertModels(t testing.TB, exp Models, act Models) {
	t.Helper()

	if len(exp) != len(act) {
		t.Fatalf("expected %d models, got %d", len(exp), len(act))
	}

	exps := fmt.Sprintf("%#v", exp)
	acts := fmt.Sprintf("%#v", act)

	if acts != exps {
		t.Fatalf("expected %s, got %s", exps, acts)
	}
}
```

**Review by markbates — COMMENTED** [review]
Arun, excellent work overall. 

I've made some suggestions, please apply them to all of your code.

File: week05/errors_test.go (hunk 4 above)
Comment (arunpariyar) [reply week05/errors_test.go:185]: I have made the changes accordingly and also consolidated all the ErrTableNotFound to table driven tests, Thank you for the feedback Mark and hope you will find changes satisfactory

File: week05/store_test.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	}{
+		{name: "store with orders",
+			store: &Store{
+				data: data{
+					"orders": {
+						{"desktop": "dell"},
+						{"mobile": "apple"},
+						{"smartwatch": "xioami"},
+						{"gps": "garmin"},
+					},
```
Comment (markbates) [inline week05/store_test.go:38]: You probably want to break out stuff like this into a variable. it'll make reading the test cases a little. Plus, you can re-use that variable for multiple test cases, which can be really useful.x

**Review by markbates — COMMENTED** [review]
Excellent work!

**Pull request merged**